File: sdk/dotnet/RexecClient.cs
using System.Net.Http.Headers;
using System.Net.Http.Json;
using System.Text.Json;
using System.Text.Json.Serialization;

namespace Rexec;

/// <summary>
/// Main client for interacting with Rexec API.
/// </summary>
/// <example>
/// <code>
/// var client = new RexecClient("https://your-instance.com", "your-token");
/// var container = await client.Containers.CreateAsync("ubuntu:24.04");
/// await client.Containers.StartAsync(container.Id);
/// </code>
/// </example>
public class RexecClient : IDisposable
{
    private readonly HttpClient _httpClient;
    private readonly string _baseUrl;
    private readonly string _token;
    private readonly JsonSerializerOptions _jsonOptions;
    private bool _disposed;

    /// <summary>
    /// Container management service.
    /// </summary>
    public ContainerService Containers { get; }

    /// <summary>
    /// File operations service.
    /// </summary>
    public FileService Files { get; }

    /// <summary>
    /// Terminal connection service.
    /// </summary>
    public TerminalService Terminal { get; }

    /// <summary>
    /// Create a new Rexec client.
    /// </summary>
    /// <param name="baseUrl">Base URL of your Rexec instance</param>
    /// <param name="token">API token for authentication</param>
    public RexecClient(string baseUrl, string token)
    {
        _baseUrl = baseUrl.TrimEnd('/');
        _token = token;

        _jsonOptions = new JsonSerializerOptions
        {
            PropertyNamingPolicy = JsonNamingPolicy.CamelCase,
            DefaultIgnoreCondition = JsonIgnoreCondition.WhenWritingNull
        };

        _httpClient = new HttpClient
        {
            BaseAddress = new Uri(_baseUrl),
            Timeout = TimeSpan.FromSeconds(30)
        };
        _httpClient.DefaultRequestHeaders.Authorization = new AuthenticationHeaderValue("Bearer", token);
        _httpClient.DefaultRequestHeaders.Accept.Add(new MediaTypeWithQualityHeaderValue("application/json"));

        Containers = new ContainerService(this);
        Files = new FileService(this);
        Terminal = new TerminalService(this);
    }

    internal string BaseUrl => _baseUrl;
    internal string Token => _token;
    internal JsonSerializerOptions JsonOptions => _jsonOptions;

    internal string GetWebSocketUrl(string path)
    {
        var uri = new Uri(_baseUrl);
        var wsScheme = uri.Scheme == "https" ? "wss" : "ws";
        var port = uri.Port != -1 ? uri.Port : (uri.Scheme == "https" ? 443 : 80);
        return $"{wsScheme}://{uri.Host}:{port}{path}";
    }

    internal async Task<T?> RequestAsync<T>(HttpMethod method, string path, object? body = null, CancellationToken cancellationToken = default)
    {
        var request = new HttpRequestMessage(method, path);

        if (body != null)
        {
            request.Content = JsonContent.Create(body, options: _jsonOptions);
        }

        var response = await _httpClient.SendAsync(request, cancellationToken);

        if (!response.IsSuccessStatusCode)
        {
            var errorBody = await response.Content.ReadAsStringAsync(cancellationToken);
            var errorMessage = ExtractErrorMessage(errorBody);
            throw new RexecException((int)response.StatusCode, errorMessage);
        }

        if (typeof(T) == typeof(object) || response.Content.Headers.ContentLength == 0)
        {
            return default;
        }

        return await response.Content.ReadFromJsonAsync<T>(_jsonOptions, cancellationToken);
    }

    internal async Task<byte[]> RequestBytesAsync(HttpMethod method, string path, CancellationToken cancellationToken = default)
    {
        var request = new HttpRequestMessage(method, path);
        var response = await _httpClient.SendAsync(request, cancellationToken);

        if (!response.IsSuccessStatusCode)
        {
            throw new RexecException((int)response.StatusCode, "Download failed");
        }

        return await response.Content.ReadAsByteArrayAsync(cancellationToken);
    }

    private string ExtractErrorMessage(string body)
    {
        try
        {
            var error = JsonSerializer.Deserialize<ErrorResponse>(body, _jsonOptions);
            if (!string.IsNullOrEmpty(error?.Error))
            {
                return error.Error;
            }
        }
        catch { }

        return string.IsNullOrEmpty(body) ? "Unknown error" : body;
    }

    public void Dispose()
    {
        if (!_disposed)
        {
            _httpClient.Dispose();
            _disposed = true;
        }
        GC.SuppressFinalize(this);
    }

    private class ErrorResponse
    {
        public string? Error { get; set; }
        public string? Message { get; set; }
    }
}

File: sdk/php/src/RexecClient.php
<?php

declare(strict_types=1);

namespace Rexec;

use GuzzleHttp\Client as HttpClient;
use GuzzleHttp\Exception\GuzzleException;
use GuzzleHttp\Exception\RequestException;

/**
 * Main client for interacting with the Rexec API.
 *
 * @example
 * $client = new RexecClient('https://rexec.sh', $token);
 * $list = $client->containers()->list();
 * $container = $client->containers()->create('ubuntu', ['name' => 'demo']);
 * $client->containers()->delete($container->id);
 */
class RexecClient
{
    private HttpClient $httpClient;
    private string $baseUrl;
    private string $token;

    private ContainerService $containers;
    private FileService $files;
    private TerminalService $terminal;

    /**
     * @param string $baseUrl Base URL of your Rexec instance
     * @param string $token   API token (Bearer)
     * @param array  $options Additional Guzzle options
     */
    public function __construct(string $baseUrl, string $token, array $options = [])
    {
        $this->baseUrl = rtrim($baseUrl, '/');
        $this->token = $token;

        $defaultOptions = [
            'base_uri' => $this->baseUrl . '/',
            'timeout' => 30,
            'http_errors' => false,
            'headers' => [
                'Authorization' => 'Bearer ' . $token,
                'Accept' => 'application/json',
                'Content-Type' => 'application/json',
                'User-Agent' => 'pipeops-rexec-php/1.0.1',
            ],
        ];

        $this->httpClient = new HttpClient(array_merge($defaultOptions, $options));

        $this->containers = new ContainerService($this);
        $this->files = new FileService($this);
        $this->terminal = new TerminalService($this);
    }

    public function containers(): ContainerService
    {
        return $this->containers;
    }

    public function files(): FileService
    {
        return $this->files;
    }

    public function terminal(): TerminalService
    {
        return $this->terminal;
    }

    public function getBaseUrl(): string
    {
        return $this->baseUrl;
    }

    public function getToken(): string
    {
        return $this->token;
    }

    /**
     * Build a WebSocket URL for a path (e.g. /ws/terminal/{id}).
     */
    public function getWebSocketUrl(string $path): string
    {
        $parsed = parse_url($this->baseUrl);
        $scheme = ($parsed['scheme'] ?? 'http') === 'https' ? 'wss' : 'ws';
        $host = $parsed['host'] ?? 'localhost';
        $port = $parsed['port'] ?? null;
        $portPart = $port !== null ? ':' . $port : '';
        if ($path === '' || $path[0] !== '/') {
            $path = '/' . $path;
        }

        return "{$scheme}://{$host}{$portPart}{$path}";
    }

    /**
     * Make a JSON API request.
     *
     * @throws RexecException
     */
    public function request(string $method, string $path, ?array $body = null): mixed
    {
        try {
            $options = [];
            if ($body !== null) {
                $options['json'] = $body;
            }

            $response = $this->httpClient->request($method, ltrim($path, '/'), $options);
            $status = $response->getStatusCode();
            $content = (string) $response->getBody();

            if ($status < 200 || $status >= 300) {
                throw RexecException::fromResponse($status, $content);
            }

            if ($content === '') {
                return null;
            }

            $decoded = json_decode($content, true);
            if (json_last_error() !== JSON_ERROR_NONE) {
                throw new RexecException('Invalid JSON response: ' . json_last_error_msg(), $status);
            }

            return $decoded;
        } catch (RexecException $e) {
            throw $e;
        } catch (RequestException $e) {
            $status = $e->getResponse()?->getStatusCode() ?? 0;
            $body = $e->getResponse() ? (string) $e->getResponse()->getBody() : $e->getMessage();
            throw RexecException::fromResponse($status, $body, $e);
        } catch (GuzzleException $e) {
            throw new RexecException($e->getMessage(), 0, $e);
        }
    }

    /**
     * Make a request and return raw response bytes/string.
     *
     * @throws RexecException
     */
    public function requestBytes(string $method, string $path): string
    {
        try {
            $response = $this->httpClient->request($method, ltrim($path, '/'));
            $status = $response->getStatusCode();
            $content = (string) $response->getBody();

            if ($status < 200 || $status >= 300) {
                throw RexecException::fromResponse($status, $content);
            }

            return $content;
        } catch (RexecException $e) {
            throw $e;
        } catch (GuzzleException $e) {
            throw new RexecException($e->getMessage(), 0, $e);
        }
    }
}
